CRM/Admin/Form/Options.php

```diff
@@ -376,10 +376,36 @@ public static function formRule($fields, $files, $self) {
       }
     }
 
+    $dataType = self::getOptionGroupDataType($self->_gName);
+    if ($dataType) {
+      $validate = CRM_Utils_Type::validate($fields['value'], $dataType, FALSE);
+      if (!$validate) {
+        CRM_Core_Session::setStatus(
+          ts('Data Type of the value field for this option value does not match ' . $dataType),
+          ts('Value field Data Type mismatch'));
+      }
+    }
     return $errors;
   }
 
   /**
+   * Get the DataType for a specified Option Group.
+   *
+   * @param string $optionGroupName name of the option group
+   *
+   * @return string|null
+   */
+  public static function getOptionGroupDataType($optionGroupName) {
+    $optionGroup = civicrm_api3('OptionGroup', 'get', array(
+      'sequential' => 1,
+      'name' => $optionGroupName,
+    ));
```

Comment:
Can we use `getFieldValue()` directly here to get the group id instead of an API call ? Something like -

    $optionGroupId = CRM_Core_DAO::getFieldValue('CRM_Core_DAO_OptionGroup', $optionGroupName, 'id', 'name');